source/Calamari/Kubernetes/Commands/Executors/KustomizeExecutor.cs

```diff
@@ -0,0 +1,92 @@
+#if !NET40
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Calamari.Common.Commands;
+using Calamari.Common.Plumbing.Extensions;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Kubernetes.Integration;
+using Calamari.Kubernetes.ResourceStatus.Resources;
+using Newtonsoft.Json.Linq;
+
+namespace Calamari.Kubernetes.Commands.Executors
+{
+    public interface IKustomizeKubernetesApplyExecutor : IKubernetesApplyExecutor
+    {
+    }
+    
+    class KustomizeExecutor : BaseKubernetesApplyExecutor, IKustomizeKubernetesApplyExecutor
+    {
+        const int MinimumKubectlVersionMajor = 1;
+        const int MinimumKubectlVersionMinor = 24;
+        readonly ILog log;
+        readonly Kubectl kubectl;
+
+        public KustomizeExecutor(ILog log, Kubectl kubectl) : base(log, kubectl)
+        {
+            this.log = log;
+            this.kubectl = kubectl;
+        }
+
+        protected override IEnumerable<ResourceIdentifier> ApplyAndGetResourceIdentifiers(RunningDeployment deployment)
+        {
+            var variables = deployment.Variables;
+            var overlayPath = variables.Get(SpecialVariables.KustomizeOverlayPath);
+
+            if (overlayPath == null)
+            {
+                log.Warn("Kustomization directory not specified");
+                return Enumerable.Empty<ResourceIdentifier>();
+            }
```

Comment:
Do you know what happens in the step package kustomize action if the overlay path isn't set? In this version the deployment will succeed but with warnings and that doesn't feel like correct behaviour to me. We will need to add some validation to the frontend to prevent the overlay path not been set but I think this should throw an exception might be worth double checking with others in the team though.

![Screenshot 2024-02-02 at 10 01 22 am](https://github.com/OctopusDeploy/Calamari/assets/3272176/f8fcb562-0ffd-4a05-a78c-6f231c709bf4)

Reply:
Good question. My thought was that this is already being validated by the [frontend](https://github.com/OctopusDeploy/OctopusDeploy/pull/22446/files#diff-942a70e020f19d3d7fd87a3be24c8920068a26aa80da2fa2b1ac2046bab78099R14) such that it is a required field. From Calamari's perspective, if the overlay path is not specified (not possible unless the user runs the Calamari command directly somehow) then it would be a no-op but not a fatal exception. But at the same time I guess you could argue that if you ran the `kubectl apply -k` command with no directory (or it fails to find the kustomization file) then it results in an error rather than no-op. So I'm more than happy to make it an exception just to be consistent.

Reply:
I think keeping it consistent and adding the exception or removing the check for a null overlay path and relying entirely the error from `kubectl` is a good path forward so we avoid representing a deployment as successful when nothing was actually deployed.

I have a TODO to double check the frontend validation, I think it just displays an error but doesn't prevent you from saving process. Probably need to add a new kind of validator something like the [raw yaml validation](https://github.com/OctopusDeploy/OctopusDeploy/blob/c0e8d0a3eb6a08582a39a52dd9fc68aaa819624a/source/Octopus.Kubernetes/Validation/KubernetesDeployRawYamlActionValidator.cs#L14).